### PR TITLE
Fixed #27724 -- SelectDateWidget clears date and month if year is not selected

### PR DESCRIPTION
**Pull request by aaboffill**
Fixed #27724 `SelectDateWidget` clears date and month if year is not selected.

If year is not selected in `SelectDateWidget`, then all the inputs are cleared as opposed to a state where date or month is missing.

### CODE REVIEW

**Comment by timgraham** [comment]
I'm not sure if testing `date_re` is the best approach. If the implementation changes, the tests may need to be rewritten. Would it be better to test `format_value()` instead?

**Comment by aaboffill** [comment]
@timgraham I followed your suggestion and I added some changes in order to use `format_value()` instead of `date_re`.

File: tests/forms_tests/widget_tests/test_selectdatewidget.py [hunk 1]
```diff
@@ -485,3 +485,39 @@ def test_value_omitted_from_data(self):
         self.assertIs(self.widget.value_omitted_from_data({'field_day': '1'}, {}, 'field'), False)
         data = {'field_day': '1', 'field_month': '12', 'field_year': '2000'}
         self.assertIs(self.widget.value_omitted_from_data(data, {}, 'field'), False)
+
+    def test_regexp_date_re(self):
+        # check valid formats
+        # 1 digit for (day & month)
+        self.assertDictEqual(self.widget.format_value('2000-1-1'), {'day': 1, 'month': 1, 'year': 2000})
```
Comment (timgraham) [inline tests/forms_tests/widget_tests/test_selectdatewidget.py]: `assertEqual` should be fine. See note in the [Python docs](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addTypeEqualityFunc), "Note that it’s usually not necessary to invoke these methods directly."

**Comment by aaboffill** [comment]
@timgraham I added a new PR [8206](https://github.com/django/django/pull/8206)  with the additional non-regressions tests. I followed all your suggestions. I will add a new cleaned PR with the solution to fix [#27724](https://code.djangoproject.com/ticket/27724) in order to avoid the wrong commit [6a2bc33](https://github.com/django/django/pull/8108/commits/6a2bc334531dfad84fb3d8ea037787067d925a3f)

**Comment by aaboffill** [comment]
@timgraham I added the new PR [8209](https://github.com/django/django/pull/8209) in order to be merged instead of this PR. This PR should be closed without merge. With that action we can avoid to include the commit [6a2bc33](https://github.com/django/django/pull/8108/commits/6a2bc334531dfad84fb3d8ea037787067d925a3f) into the history.

**Comment by felixxm** [comment]
Updated in #8209.

**Pull request closed without merging**